python/mxnet/gluon/data/sampler.py

```diff
@@ -136,3 +138,74 @@ def __len__(self):
         raise ValueError(
             "last_batch must be one of 'keep', 'discard', or 'rollover', " \
             "but got %s"%self._last_batch)
+
+
+class AliasMethodSampler(object):
+    """ The Alias Method: Efficient Sampling with Many Discrete Outcomes.
+    Can be use in NCELoss.
+
+    Parameters
+    ----------
+    K : int
+        Number of events.
+    probs : array
+        Probability of each events, corresponds to K.
+
+    References
+    -----------
+        https://hips.seas.harvard.edu/blog/2013/03/03/the-alias-method-efficient-sampling-with-many-discrete-outcomes/
+    """
```

Comment:
Since the AliasMethod is faster than the BinarySearch, we can consider to revise the implementation of SampleMultinomial https://github.com/apache/incubator-mxnet/blob/master/src/operator/random/sample_multinomial_op.h#L111-L135.

Reply:
Great. I'll try to do this once we finish this PR.

Reply:
@sxjscience do we expect to construct the table every time when `SampleMultinomial` is called?

Reply:
@eric-haibin-lin Yes, we need to construct the table every time. Our current implementation is to first construct a cumsum table and then search the columns to get the result. The search complexity of the current implementation is O(N) and the best possible search complexity if O(log(N)).  For this approach, the construction complexity is O(N) and the search complexity is O(1), which will faster when we draw multiple samples from the same multinomial distribution.